File: 3085-minimum-deletions-to-make-string-k-special/3085-minimum-deletions-to-make-string-k-special.cpp
using namespace std;

class Solution {
public:
    int minimumDeletions(const string& word, int k) {
        array<int, 26> freq{};
        for (unsigned char c : word)
            ++freq[c - 'a'];

        vector<int> freqList;
        freqList.reserve(26);
        for (int f : freq)
            if (f > 0)
                freqList.push_back(f);

        sort(begin(freqList), end(freqList));

        int n = freqList.size(), total = word.size(), removedLeft = 0,
            inWindowSum = 0, best = INT_MAX, r = 0;

        for (int l = 0; l < n; ++l) {
            int base = freqList[l];
            int limit = base + k;

            while (r < n && freqList[r] <= limit)
                inWindowSum += freqList[r++];

            int rightCount = n - r;
            int rightSum = total - inWindowSum;
            int delRight = rightSum - rightCount * limit;

            best = min(best, removedLeft + delRight);

            total -= freqList[l];
            removedLeft += freqList[l];
            inWindowSum -= freqList[l];
        }

        return best;
    }
};

File: 3085-minimum-deletions-to-make-string-k-special/3085-minimum-deletions-to-make-string-k-special.py
class Solution(object):
    def minimumDeletions(self, word, k):
        freq = [0] * 26
        for ch in word:
            freq[ord(ch) - ord('a')] += 1

        freq_list = [f for f in freq if f > 0]
        freq_list.sort()

        n = len(freq_list)
        total = len(word)
        removed_left = 0
        window_sum = 0
        result = total  
        r = 0

        for l in range(n):
            base = freq_list[l]
            limit = base + k

            while r < n and freq_list[r] <= limit:
                window_sum += freq_list[r]
                r += 1
            
            right_count = n - r
            right_sum = total - window_sum
            del_right = right_sum - right_count * limit

            curr = removed_left + del_right
            if curr < result:
                result = curr

            total -= base
            removed_left += base
            window_sum -= base

        return result